bocs_control/data_writer.py

```diff
@@ -0,0 +1,84 @@
+"""=============================================================================
+DataWriter CLASS
+--------------------------------------------------------------------------------
+
+============================================================================="""
+from datetime import datetime as dt
+import logging
+import re
+import threading
+
+import bocs_control.config as cfg
+
+# ===============================================================================
+class DataWriter(threading.Thread):
+    """
+    DataWriter describes a thread whose purpose is to flush lines of serial
+    input from a shared FIFO queue into different files, depending on the nature
+    of the serial input to be flushed.
+    """
+
+    def __init__(self, shared_queue):
+        threading.Thread.__init__(self)
+        logging.info("Initialising data logging thread")
+        self.queue = shared_queue
+
+    def dequeue_data(self):
+        """
+        Pull a line of data from the front of a shared queue.
+        """
+        logging.debug("Dequeueing data")
+        line = self.queue.get(block=True, timeout=None)
+        logging.debug(f"Queue size is now {self.queue.qsize()}")
+
+        return line
+
+    def write_data(self, data):
+        """
+        Write data to the appropriate log file, named by instrument name and
+        date.
+        """
+        logging.debug("Writing data to log file")
+        data_fields = data.split(",")
+
+        if re.match("ERROR", data_fields[1]):
+            logging.error(
+                f"Error in received transmission. Check error log ({cfg.ERROR_LOG_FN}) for further details"
+            )
+            with open(cfg.ERROR_LOG_FN, "a") as error_log:
+                error_log.write(data_fields[1])
+            return
+
+        if data_fields[0] not in cfg.INSTRUMENTS:
+            logging.error(
+                f"{data_fields[0]} isn't a recognised instrument identifier, taking timestamp from Pi clock"
+            )
+            date = dt.now()
+        else:
+            try:
+                date = dt.utcfromtimestamp(int(data_fields[1]))
+            except ValueError:
+                # This occurs every 2s when whitespace is available in the port
+                # and can be safely ignored
```

Comment:
Can you explain this to me?

Reply:
This is leftover from the original code. After a successful line read, a whitespace character is next available in the serial port, which this except block ignores. I added this comment just to explain that this scenario occurs often and we aren't discarding data.

Reply:
`timeout=None` seems preferable - I'll make that change